### ISSUE TITLE
Lazy loading of services

### ISSUE THREAD

**Issue by dimkir** [avighier/simple-dijs#54]
Currently if I have a big application, which for example uses AMD modules and RequireJS, my dependency container would have to depend on all of the services within the application to work, e.g.: 

```javascript
define('myContainer', [
'./Di',
'export',
 './modules/A',
 './modules/B',
'./modules/C',
'./modules/D',
'./modules/E',
'./modules/F',
'./modules/G',
], function(Di, export,  A,B,C,D,E,F,G){
  
   var di= new Di();

  // first 
   di.set('a', function(di){
       return new A();
   });

  di.set('b', function(di){
     return new B(di.get('a'));
  });

 // second
  di.set('c', function(){
      return new C();
  });
  di.set('d', function(di){
     return new D(di.get('c'));
  });

  // third
  di.set('e', function(){
     return new E();
  });

  di.set('f', function(){
     return new F();
  });
  
  di.set('g', function(di){
     return new G(
          di.get('f)',
          di.get('e')
     );
  });

   export = di;
});

```

It is easy to see that there's a lot of services which do NOT depend on anyone else, and even dependent modules fall into three separate sections. 

Upon starting up my app, it may be only using `first` section of services, and if user never used some feature, it may never have gone to use `third` section of services.  However upon loading of the DI container, they're all need to be loaded into memory, even though they're not going to be used straight away.

Is there a way to create a "lazy" service loading. So that we can do something like that:

```javascript
  di.setLazy('lazy-Z', function(di, callback){
        require('./modules/z', function(z){
               var z = new Z(
                     di.get('a'),
                     di.get('b')
                );
                callback(z);
        });
   });
```

**Comment by avighier** [comment]
Hi,

Yes, it's a shame to load all the dependencies to register them into the container.

So, the require should be inside the service definition, like your example.

Your example uses a callback, and it's necessary to Di component to inject a callback. I use promises, but probably a lot of people use callbacks.

I will think about your problem and your solution (like injecting callback into di.set()). Do that is more pollution for the code where promises doesn't, however I will think about efficient and elegant way.

If your problem is urgent, I invite you to fork and implement a solution.

Nice day,

**Comment by avighier** [comment]
I think a first solution :

- di.set() :

di.set('x', function (di) { return x; });  // Registering sync
di.set('x', function (di, callback) { callback(x); })  // Registering async, possible name : di.setAsync

- di.get()

di.get('x')        for sync only
di.get('x', function (x) {});      for async (and sync, why not)

Possible add automatic callback<->Promise translation, but not the priority.

**Comment by avighier** [comment]
Others possibles solutions :

```
di.set('x', function (di) {
   return di.async(function (callback) {
         callback(x)
   });
});
```

**Comment by dimkir** [comment]
@avighier Thank you for your quick reply! 

I actually didn't know that it is possible to return promises... Now that I look again at the readme on the main page, I can see at the bottom example with promises. But I guess I overlooked it earlier, because it doesn't actually show how to instantiate a class or dependency asynchronously and is very short.

When I submitted the question, I was interested in having an async way to define a service (not obligatory callback based, promises would do just fine).  So I will try to experiment with the promise way now...

**Comment by avighier** [comment]
Hi,

Promise is a normal object, if you return it in the set(), you will give it in the get(). I knew more WhenJs than native Promise (that is less powerfull, less error-sure, etc). But a day I answer to me if native Promise should have good behavior for multiple get() call on the same container identifier (and so, multiple then() call on the same promise), and I have written a test for that. All was Okay, but I have not kept the test because it was out of subject (or maybe not but it was my reflection).

I am happy that Promises can satisfy you. However, callback are the rule for Node API, and for a lot of library/people, whatever what I think. I think about an "official" async mecanism for simple-dijs, that allow using Promise and/or Callback easily. I just need "today" to write API and behaviors tests to know the direction.

For your case, so, if you are not stuck, it's the main point.

**Comment by avighier** [comment]
I would like something easy. I think about two things :

```javascript
        di.set('id', function (di, callback) {
            fs.readFile('config', callback);
        });

        di.set('id', function () {
            return di.async(function (callback) {
                fs.readFile('config', callback);
            })
        });
```

I think the first is the more easy and simple for people, but more "magic", because depends on the presence or not of a second argument. The second is more easy and clear to implements because the idea is that di.async() will returns a Promise, so the implementation of di.set will not change. I have also thank about di.setAsync(), but it will add complexity (but it's the less magic), it will need a di.getAsync(), and I think a only one set() and get() is more simple.

The idea is that the get() will returns a Promise in case of Async and allow a second parameter callback. My idea is to allow using Promise and callback easily, and something like :

```javascript
   di.set('id', function (id, callback) { callback(null, config); });
   di.get('id').then(function (config) {});
```

And the opposite (set with promise and get with callback), posible.

I would like to have opinions (by a "callback user") about the set(), the first or the second ?

PS : the first is used by Mocha for example to determine if an it() is async or sync.

Thanks.

**Comment by avighier** [comment]
Added commit that references this issue 8d25d14

**Comment by avighier** [comment]
Published v1.7.0 with this feature
